### ISSUE TITLE
Missing capture formats from GetVideoCaptureFormatsAsync on HL2

### ISSUE THREAD

**Issue by eirikhollis** [microsoft/MixedReality-WebRTC#159]
On HL2, when calling `PeerConnection.GetVideoCaptureDevicesAsync()`, the only capture device found is the "QC Back Camera".
If the capture device is inserted into `PeerConnection.GetVideoCaptureFormatsAsync(deviceId)`, the only resolutions returned are (W x H x FPS)
```
2272 x 1278 x 30.0000019073486
896 x 504 x 30.0000019073486
2272 x 1278 x 15.0000009536743
896 x 504 x 15.0000009536743
```

Forcing the capture resolution to 720p does indeed work, which indicates that there are missing capture formats returned from `GetVideoCaptureFormatsAsync`.

**Comment by djee-ms** [comment]
The problem is not missing formats, the problem is that either we are missing an argument to specify the video profile, or we should return a collection of all available video profiles and their associated video capture formats. Here the list you get is the list of video capture formats for the default video profile.

**Comment by djee-ms** [comment]
As a **temporary** workaround (and please keep it temporary, because this is an implementation detail, so I won't guarantee this will always work) you can directly call methods of the UWP `MediaCapture` API like [`MediaCapture.FindKnownVideoProfiles()`](https://docs.microsoft.com/en-us/uwp/api/windows.media.capture.mediacapture.findknownvideoprofiles) to list the profiles, and then access [`MediaCaptureVideoProfile.SupportedRecordMedia`](https://docs.microsoft.com/en-us/uwp/api/windows.media.capture.mediacapturevideoprofile.supportedrecordmediadescription) to get the capture formats.

**Comment by eirikhollis** [comment]
Ah, I understand. Thanks for the suggestion for the temporary workaround.

**Comment by eirikhollis** [comment]
As a suggestion, I would probably have video profile as an optional parameter, such as:
```
GetVideoCaptureFormatsAsync(string deviceId, VideoProfileKind videoProfile = 0);
```
where if you do not specify a video profile (Unspecified), it will return all formats and their associated video profile, but if you set a specific profile, then it will only return the formats related to that profile.

**Comment by djee-ms** [comment]
I had a quick look at it and it's not a trivial change because although we expose the profile ID for capture, the enumeration is still done generically via the Google API which doesn't have any concept of video profile. So this requires some change to transition to explicit Media Foundation based format enumeration

**Comment by eirikhollis** [comment]
Note: Not sure if this is related to this or a new issue, let me know what you think.

In Unity (x64), when calling `GetVideoCaptureFormatsAsync`, only formats with `framerate = 30` are returned, but when deploying to UWP (x86) and running on local computer with the same webcamera, all framerates such as `7, 10, 15, 20, 30` are returned.

Any reason why this happens?

**Comment by djee-ms** [comment]
Yes, Desktop and UWP use completely different implementations (DirectShow and MediaFoundation, respectively) and enumerate devices and their formats via completely different Windows APIs, which usually return different results as some older devices may not be registered with MF while some newer ones (networked webcams, etc.) may not be registered with DS. This all depends on the device and its driver. In addition DS doesn't have a concept of video profile, so lists all capture formats, while MF as explained above will list only the capture formats for the currently selected video profile (or the default one). On HoloLens 2 for example, MF will list only a single resolution with the default profile, whereas it can capture with many more, but they are only exposed via other video profiles than the default.

**Comment by djee-ms** [comment]
_Note:_ The WebRTC UWP team is working on adding a unified video module for Desktop and UWP to the Google upstream repository : https://webrtc-review.googlesource.com/c/src/+/167021. This will prevent this kind of discrepancies between the Unity Editor (Desktop) and some UWP device.

**Comment by eirikhollis** [comment]
Nice, thanks for the quick reply :+1:

**Comment by michalek-marcin** [comment]
Hi @djee-ms , I have same problem as @eirikhollis video-call works in Unity Editor, but on HL2 the only capture device found is the "QC Back Camera" and I have no picture or sound.

The question is: If I'm using Unity components like PeerConnection, LocalVideoSource, LocalAudioSource etc, should it work on HL1/HL2 device? Or should I have two implementations: one for Unity editor and one for UWP device?

**Comment by djee-ms** [comment]
@michalek-marcin the Unity integration abstracts the platform, it works both in the Editor and on HoloLens 1/2. The only notable difference is that H.264 support only works with UWP, so if you use the Unity editor you need to use another encoding (_e.g._ VP8).

I am not sure what you mean by:

> on HL2 the only capture device found is the "QC Back Camera"

...because this is the only camera of the device, so there's nothing wrong. What did you expect here? 

This bug is about some capture formats missing from the overall list returned by `GetVideoCaptureFormatAsync()`. But the list is never empty anyway, so this is not about some issue with picture or sound. If you have neither video nor audio then the connection is likely not established at all.

**Comment by michalek-marcin** [comment]
@djee-ms Sorry, I expected something like "Front camera".

I asked about unity integration and if it works on UWP device, because on device I didn't have a picture and sound. My connection was never established as you said. After read your comment I found place where PeerConnection initialization throws unhandled exception:

`var accessTask = mediaAccessRequester.InitializeAsync(mediaSettings).AsTask(token);`

At application start I already ask user to give access to camera (Qr code scan), so I skip this and now everything works! :)

**Comment by eirikhollis** [comment]
@michalek-marcin Does that mean that according to your findings, the `mediaAccessRequester.InitializeAsync` function throws an error if video/microphone access has already been given to the application?

**Comment by michalek-marcin** [comment]
@eirikhollis I'm not 100% sure, but in my case when I comment code which ask for camera access everything started working.

In my app user must scan qr code to log in, so I already have access to camera. But still I have edge case: user can go to HL menu, go to settings and revoke permission to access to camera. Right now I must implement other features, so I must leave it as it is right now, but I will investigate this in the near future.

**Comment by djee-ms** [comment]
This sounds similar to a bug we saw when combining MixedReality-WebRTC and MixedReality-SpectatorView. See #175 and microsoft/MixedReality-SpectatorView#322 for details. Normally if the app is already authorized then `mediaAccessRequester.InitializeAsync()` silently returns, but does not throw any exception.

**Comment by djee-ms** [comment]
@eirikhollis this should be fixed now, and will be in the 2.0.0 release.
The API has changed to:
- Add some function to enumerate video profiles
- Take a video profile ID or kind as input to capture format enumeration
